File: app/src/main/java/seigneur/gauvain/mycourt/ui/shotDraft/ShotDraftViewModel.java
package seigneur.gauvain.mycourt.ui.shotDraft;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Maybe;
import io.reactivex.disposables.CompositeDisposable;
import seigneur.gauvain.mycourt.data.model.Draft;
import seigneur.gauvain.mycourt.data.repository.ShotDraftRepository;
import seigneur.gauvain.mycourt.data.repository.TempDataRepository;
import seigneur.gauvain.mycourt.utils.Constants;
import seigneur.gauvain.mycourt.utils.SingleLiveEvent;
import timber.log.Timber;

public class ShotDraftViewModel extends ViewModel {

    @Inject
    ShotDraftRepository mShotDraftRepository;

    @Inject
    TempDataRepository mTempDataRepository;

    private CompositeDisposable mCompositeDisposable = new CompositeDisposable();

    private final SingleLiveEvent<Void> mStopRefreshEvent = new SingleLiveEvent<>();

    private final SingleLiveEvent<Draft> mItemClicked = new SingleLiveEvent<>();

    private final MutableLiveData<List<Draft>> mShotDrafts = new MutableLiveData<>();

    private boolean isRefreshing;

    @Inject
    public ShotDraftViewModel() {}

    @Override
    public void onCleared(){
        super.onCleared();
        mCompositeDisposable.clear();
    }

    public LiveData<List<Draft>> getDrafts() {
        return mShotDrafts;
    }

    public SingleLiveEvent<Void> dbChanged() {
        return mShotDraftRepository.onDraftDBChanged;
    }

    public SingleLiveEvent<Draft> getItemClickedEvent() {
        return mItemClicked;
    }


    public void fetchShotDrafts() {
        isRefreshing=false;
        mCompositeDisposable.add(fetchDrafts()
                .subscribe(
                        this::doOnDraftFound,
                        this::doOnError,
                        this::doOnNothingFound
                )
        );
    }

    public void onRefresh(boolean fromSwipeRefresh) {
        if (fromSwipeRefresh) {
            isRefreshing=true;
            mCompositeDisposable.add(fetchDrafts()
                    .subscribe(
                            this::doOnDraftFound,
                            this::doOnError,
                            this::doOnNothingFound));
        } else {
            Timber.d("not refreshing, nothing happened");
        }
    }

    /**
     * get ShotDrafts list from DB - Use mayBe because the list will be small
     * @return - List of ShotDraft
     */
    private Maybe<List<Draft>> fetchDrafts() {
        Timber.d ("getPostFromDB called");
        return mShotDraftRepository.getShotDraft();
    }

    /**
     * ShotDrafts being found in DB - do something with it
     * @param shotDrafts - list Found in DB
     */
    private void doOnDraftFound(List<Draft> shotDrafts){
        Timber.d("list loaded"+shotDrafts.toString());
        //todo - live data
        mShotDrafts.setValue(shotDrafts);

        /*if (!shotDrafts.isEmpty()) {
        } else {

        }*/

    }

    /**
     * When nothing found in DB, stop refreshing and set up a dedicated view
     */
    private void doOnNothingFound(){
        if (isRefreshing) {
            mStopRefreshEvent.call();
            //mShotDraftView.stopRefresh(); //SINGLE EVENT ?
        }

    }

    /**
     * Error happened during shotDraft fetching
     * @param throwable - error
     */
    private void doOnError(Throwable throwable){
        Timber.e(throwable);
        //TODO -SINGLE EVENT?
    }


    public void onShotDraftClicked(Draft shotDraft, int position) {
        mTempDataRepository.setDraftCallingSource(Constants.SOURCE_DRAFT);
        mTempDataRepository.setShotDraft(shotDraft);
        mItemClicked.setValue(shotDraft);

        // mShotDraftView.goToShotEdition(); //TODO -SINGLE EVENT

    }

    /*
     *********************************************************************************************
     * PUBLIC METHODS CALLED IN PRESENTER
     *********************************************************************************************/


}

File: app/src/main/java/seigneur/gauvain/mycourt/ui/shotDraft/ShotDraftViewModel.kt
package seigneur.gauvain.mycourt.ui.shotDraft

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import io.reactivex.Completable

import javax.inject.Inject

import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import seigneur.gauvain.mycourt.data.model.Draft
import seigneur.gauvain.mycourt.data.repository.ShotDraftRepository
import seigneur.gauvain.mycourt.data.repository.TempDataRepository
import seigneur.gauvain.mycourt.utils.Constants
import seigneur.gauvain.mycourt.utils.SingleLiveEvent
import timber.log.Timber

class ShotDraftViewModel

@Inject
constructor() : ViewModel() {

    @Inject
    lateinit var mShotDraftRepository: ShotDraftRepository

    @Inject
    lateinit var mTempDataRepository: TempDataRepository

    private val mCompositeDisposable = CompositeDisposable()
    private val mStopRefreshEvent = SingleLiveEvent<Void>()
    private val mShotDrafts = MutableLiveData<List<Draft>>()
    private var isRefreshing: Boolean = false
    private val mDeleteOpeDone = SingleLiveEvent<Int>()


    public override fun onCleared() {
        super.onCleared()
        mCompositeDisposable.clear()
    }

    /*
    ************************************************************************************
    *  Live data and and events to be subscribed
    ************************************************************************************/
    fun dbChanged(): SingleLiveEvent<Void> {
        return mShotDraftRepository.onDraftDBChanged
    }

    fun deleteCickEvent(): SingleLiveEvent<Void> {
        return mTempDataRepository.deleteSelecteListCmd
    }

    val editMode: LiveData<Int>
        get() = mTempDataRepository.editMode

    val itemClickedEvent = SingleLiveEvent<Draft>()

    val drafts: LiveData<List<Draft>>
        get() = mShotDrafts

    fun getDeleteOpeResult(): SingleLiveEvent<Int> {
        return mDeleteOpeDone
    }
    /*
    ************************************************************************************
    *  Public functions called bu subscribers
    ************************************************************************************/
    /**
     * fetch draft list from DB
     */
    fun fetchShotDrafts() {
        isRefreshing = false
        mCompositeDisposable.add(fetchDrafts()
                .subscribe(
                        this::doOnDraftFound,
                        this::doOnError,
                        this::doOnNothingFound
                )
        )
    }

    /**
     * User has clicked on Draft in the list, deal with it
     * @param shotDraft - draft object clicked
     * @param position - position of the item in the list
     */
    fun onShotDraftClicked(shotDraft: Draft, position: Int) {
        mTempDataRepository.draftCallingSource = Constants.SOURCE_DRAFT
        mTempDataRepository.shotDraft = shotDraft
        itemClickedEvent.value = shotDraft
    }

    fun deleteSelectDrafts(ids: ArrayList<Long>) {
        for ((i, element) in ids.withIndex()) {
            val idToDelete = ids[i]
            mCompositeDisposable.add(deleteDrafts(idToDelete)
                    .subscribe(
                            this::onDeleteSucceed,
                            this::onDeleteError
                    )
            )

        }

    }

    /**
     * get ShotDrafts list from DB - Use mayBe because the list will be small
     * @return - List of ShotDraft
     */
    private fun fetchDrafts(): Maybe<List<Draft>> {
        Timber.d("getPostFromDB called")
        return mShotDraftRepository.shotDraft
    }

    private fun deleteDrafts(id:Long): Completable {
        return mShotDraftRepository.deleteDraft(id)
    }

    /**
     * ShotDrafts being found in DB - do something with it
     * @param shotDrafts - list Found in DB
     */
    private fun doOnDraftFound(shotDrafts: List<Draft>) {
        mShotDrafts.value = shotDrafts
    }

    /**
     * When nothing found in DB, stop refreshing and set up a dedicated view
     */
    private fun doOnNothingFound() {
        if (isRefreshing) {
            mStopRefreshEvent.call()
        }

    }

    /**
     * Error happened during shotDraft fetching
     * @param throwable - error
     */
    private fun doOnError(throwable: Throwable) {
        Timber.e(throwable)
        //TODO -SINGLE EVENT?
    }

    private fun onDeleteSucceed() {
       mDeleteOpeDone.value=0
       Timber.d("delete succeed")
    }

    private fun onDeleteError(throwable: Throwable) {
        Timber.e(throwable)
        mDeleteOpeDone.value=-1
    }

}
